### DISCUSSION TITLE
What is the best way to render a pixel-perfect pixel-art image?

### DISCUSSION THREAD

**Discussion by NNBnh** [hpjansson/chafa#70] [Q&A] [answered]
Hello, i have this [image](https://veekun.com/dex/media/pokemon/icons/492-sky.png) that i want to render on my terminal:

![Shaymin sky](https://veekun.com/dex/media/pokemon/icons/492-sky.png)

So i download and try to render it:

```sh
curl -sL 'https://veekun.com/dex/media/pokemon/icons/492-sky.png' > shaymin-sky.png
chafa --symbols vhalf shaymin-sky.png
```

But the result seem blurry:

![Screenshot from 2021-12-05 21-46-23](https://user-images.githubusercontent.com/43980777/144751467-e46e2281-65f5-4734-8c9c-4e0d90b980c6.png)

I want the output to be pixel-perfect, am i missing something? you guy have any suggestion?

**Comment by hpjansson** [comment]
Which terminal are you running? Try one with sixel support, or Kitty, which is supported through its own protocol. Chafa should detect your terminal and use the correct mode automatically, but if it doesn't, there's always the `-f` flag (the parameter can be `sixel`, `kitty`, `iterm`, or `symbol`.

There may still be some blurriness due to scaling. Currently, the only was around that is to prescale the image using nearest-neighbor interpolation. I.e. make the pixels 16x bigger or 32x bigger. Then run `chafa` on that.

You can do the prescaling using something like `convert in.png -filter point -resize 800% out.png`.

**Comment by acxz** [comment, marked as answer]
I spent some time trying to tackle this issue as I think I got a reasonable solution:
I found that `-filter` did some blurring so I wanted to avoid that option and just use [`-scale`](https://imagemagick.org/script/command-line-options.php?#scale) for exact pixel replication.

For the following image:
![not-shiny:shaymin-sky+male-small](https://user-images.githubusercontent.com/17132214/184521756-64f0fc03-b75f-4aa0-9919-2fb1999cf4ac.png)

Running just `chafa`:
![chafa-og-shaymin-sky](https://user-images.githubusercontent.com/17132214/184522137-a7c534d7-a015-4eba-b556-38cee9313cd6.png)

Running:
```
convert in.png -scale 800% out.png
chafa --scale 1 out.png
```

We get the pixel perfect:
![chafa-scale-shaymin-sky](https://user-images.githubusercontent.com/17132214/184522263-b2743905-9eca-478c-855c-5e5740e2ea7b.png)


Note for some small images this doesn't work so well, but for most sprites the above will serve you well.

Here is an example of another image with the prescaling done for chafa and comparing it with timg's output:
![not-shiny:pichu+female-small](https://user-images.githubusercontent.com/17132214/184522351-1511acdc-0fa3-44bd-a6d4-4d5dd259b665.png)


Chafa w/ prescaling:
![chafa-scale-pichu](https://user-images.githubusercontent.com/17132214/184522304-3e6e8985-8724-40ff-aa86-0cd42c740abe.png)

Timg's output with original file:
![timg-pichu](https://user-images.githubusercontent.com/17132214/184522331-5ad98843-1615-4979-bcb4-8b793e9b5692.png)

I would love if there was a mode for pixel perfect display in chafa that somehow bypasses the scaling of the dimension into cells. Note looking at the pipeline end to end, the prescaling does cause that pipeline to be slower than timg's pipeline, while also not being as accurate.

**Reply by hpjansson** [reply]
Great answer. You might get a little bit closer and skip ImageMagick by doing this:

```
chafa --symbols vhalf -w 1 --color-extractor median -s 15x10 --stretch pika.png
```

Quick explanation of the switches follows:

* `--symbols vhalf`: Use U+2584 only, which I think is what timg is doing.
* `-w 1`: Work as quickly as possible. This is a hack to disable smooth scaling.
* `--color-extractor median`: When colorizing the symbols, pick the median of the underlying pixels instead of the average.
* `-s 15x10`: Scale to a cell size corresponding to 1x2 pixels per cell. The dimensions come from e.g. `file pika.png`, with vertical size divided by two, rounding up.
* `--stretch`: Ignore aspect so we get the exact size specified by `-s`.

The result is almost perfect, but there's one more hurdle we can't overcome yet: the input image is 15x19 pixels, and 19 isn't evenly divisible by two. What we really want is to specify the scale in half-cells (and just pad the first or last row), but Chafa only allows integer cells. That can be fixed, though.

I think this shows there's a pixel-perfect mode in Chafa struggling to get out :-) I'll try to release it as time allows.

**Reply by acxz** [reply]
@hpjansson trying to get back into some terminal shenanigans and I remember this was a small lingering nit I had.

I was wondering if half-cell support has been added to Chafa or if there is a plan to add pixel-perfect/half cell support? If there is a plan should an issue be made?

**Reply by hpjansson** [reply]
There is pixel perfect processing since version 1.14, but only for pixel-based protocols like sixels and iterm2. I still need to make a few adjustments for it to work effortlessly with symbols. You should get the best possible half-cell output this way now:

`chafa --scale 8 --symbols vhalf -w 1`

The `--scale` option was added in version 1.12. This respects the view size, so if you want results that can be bigger than the terminal, you should use `--view-size 9999x9999` or some other suitably large extent (1.14 onwards).

**Reply by acxz** [reply]
@hpjansson I want to follow up with this as `--symbols vhalf` makes the pixel perfect output worse. See the below image for a comparison of chafa (with and without `--symbols vhalf`) and timg (which still has better output):

![image](https://github.com/user-attachments/assets/e327018f-3476-4126-926d-2a8e5dd362f0)

Is there a better symbols group I can use, one that has the same group as timg's default (quarter)? See [Half block and quarter block rendering](https://github.com/hzeller/timg?tab=readme-ov-file#half-block-and-quarter-block-rendering) on timg's readme for more info.

Edit: we do have `--symbols quad`, but that doesn't fare better (I tried with all `-w` options, and `-w 9` ends up pretty blurry)

![image](https://github.com/user-attachments/assets/8832273c-60bf-4f4f-915e-b5f11d6461f2)

**Reply by hpjansson** [reply]
If you want all the quad permutations, you have to explicitly include half-blocks too: `--symbols half+quad`. You may also get sharper pixels using `--color-extractor median`. But in order to do this perfectly, we need to implement the following:

- Decouple nearest-neighbor resampling from the `-w` switch (or always do upsampling with nearest-neighbor).
- Align image to half-block or quad-block grid. The first example looks like an alignment issue.

**Reply by acxz** [reply]
I feel like decoupling the nearest-neighbor resampling would do the trick.
I didn't even think about the alignment tbh.

Posting output of some various combinations for posterity's sake:

![image](https://github.com/user-attachments/assets/868d4ef3-586c-4ca9-b4b9-8343b50980c9)

Based on all this as of now (1.14.2) the best way to render a pixel perfect pixelart image is simply `chafa --scale 8 --symbols all -w 1`.

**Comment by clort81** [comment]
Match input and output scale.
Then convert with --symbols vhalf (and no other).
The vertical-half symbol will be invariant, represent two square (vertically stacked) pixels, controlled via fg and bg color.
Limiting the valid character to just 'upper half' maximally aids compression
